### PR TITLE
Add nbstripout pre-commit hook

### LINKED ISSUE #427: Clean metadata from notebooks during pre-commit
### Feature description

Remove unnecessary metadata from the example notebooks prior to git commit.

### Motivation

To simplify the review process by ensuring that only material changes to the example notebooks and their output are shown.

### Possible implementation

Add the nbstripout pre-commit hook from: https://github.com/kynan/nbstripout.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by NicolaCourtier**
# Description

Add the nbstripout pre-commit hook from: https://github.com/kynan/nbstripout

## Issue reference
Fixes #427

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [ ] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

**Comment by NicolaCourtier** [comment]
Note that the failed project coverage check is simply due to the extra Changelog entry.

**Review by agriyakhetarpal — APPROVED** [review]
These changes LGTM in principle; thank you, @NicolaCourtier!

File: examples/notebooks/LG_M50_ECM/1-single-pulse-circuit-model.ipynb
Comment (BradyPlanden) [inline examples/notebooks/LG_M50_ECM/1-single-pulse-circuit-model.ipynb:1]: What is the reason for removing the -q option on the `%pip install --upgrade pip ipywidgets -q` lines?  (throughout)

File: examples/notebooks/comparing_cost_functions.ipynb
Comment (BradyPlanden) [inline examples/notebooks/comparing_cost_functions.ipynb:1]: A comment alongside the random seed fix would be helpful. Something along the lines of "We are fixing the random seed for these examples to have consistent output during development, but this does not need to be done in practice".

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -32,3 +32,9 @@ repos:
         - id: rst-backticks
         - id: rst-directive-colons
         - id: rst-inline-touching-normal
+
+  - repo: https://github.com/kynan/nbstripout
+    rev: 0.7.1
+    hooks:
+      - id: nbstripout
+        args: ['--keep-output']
```
Comment (BradyPlanden) [inline .pre-commit-config.yaml:40]: I think it's worth removing empty cells as well:
```suggestion
        args: ['--keep-output', '--drop-empty-cells']
```

**Review by BradyPlanden — APPROVED** [review]
This looks great, thanks @NicolaCourtier! The patch coverage is down due to this flaky test: https://app.codecov.io/gh/pybop-team/PyBOP/pull/428/indirect-changes which I've fixed in #411: https://github.com/pybop-team/PyBOP/blob/7bae73c60da37932851f8373cc567f8f35e8a078/tests/unit/test_optimisation.py#L427 

I've left a few comments/suggestions, otherwise LGTM.

File: examples/notebooks/LG_M50_ECM/1-single-pulse-circuit-model.ipynb
Comment (NicolaCourtier) [reply examples/notebooks/LG_M50_ECM/1-single-pulse-circuit-model.ipynb:1]: Do you mean why did I add the `-q` option? This is to silence the output from pip which is essentially metadata.

Comment (BradyPlanden) [reply examples/notebooks/LG_M50_ECM/1-single-pulse-circuit-model.ipynb:1]: Ah yes! I was viewing the diffs backwards, nice addition.

**Pull request merged**